query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,384 @@
+                bail!(ErrorKind::InvalidGroundConstant)
+            }
+        }
+    }
+
+    return Ok((matrix, column_types))
+}
+
+/// Application of where functions.
+impl ConjoiningClauses {
+    /// There are several kinds of functions binding variables in our Datalog:
+    /// - A set of functions like `ground`, fulltext` and `get-else` that are translated into SQL
+    ///   `VALUES`, `MATCH`, or `JOIN`, yielding bindings.
+    /// - In the future, some functions that are implemented via function calls in SQLite.
+    ///
+    /// At present we have implemented only a limited selection of functions.
+    pub fn apply_where_fn<'s>(&mut self, schema: &'s Schema, where_fn: WhereFn) -> Result<()> {
+        // Because we'll be growing the set of built-in functions, handling each differently, and
+        // ultimately allowing user-specified functions, we match on the function name first.
+        match where_fn.operator.0.as_str() {
+            "ground" => self.apply_ground(schema, where_fn),
+            _ => bail!(ErrorKind::UnknownFunction(where_fn.operator.clone())),
+        }
+    }
+
+    #[allow(unused_variables)]
+    pub fn apply_ground<'s>(&mut self, schema: &'s Schema, where_fn: WhereFn) -> Result<()> {
+        if where_fn.args.len() != 2 {
+            bail!(ErrorKind::InvalidNumberOfArguments(where_fn.operator.clone(), where_fn.args.len(), 2));
+        }
+
+        let mut args = where_fn.args.into_iter();
+
+        // TODO: process source variables.
+        match args.next().unwrap() {
+            FnArg::SrcVar(SrcVar::DefaultSrc) => {},
+            _ => bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "source variable".into(), 0)),
+        }
+
+        if where_fn.binding.is_empty() {
+            // The binding must introduce at least one bound variable.
+            bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::NoBoundVariable));
+        }
+
+        if !where_fn.binding.is_valid() {
+            // The binding must not duplicate bound variables.
+            bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::RepeatedBoundVariable));
+        }
+
+        // Turn input constant into a matrix, ready to feed into a VALUES clause.
+        let constant = args.next().unwrap();
+        let rows = { // Scope borrow of self.
+            let operator = where_fn.operator;
+            let lookup = |in_var| {
+                if !self.input_variables.contains(&in_var) {
+                    bail!(ErrorKind::UnboundVariable((*in_var.0).clone()));
+                }
+                match self.bound_value(&in_var) {
+                    Some(ref in_value) => Ok(in_value.clone()),
+                    None => bail!(ErrorKind::UnboundVariable((*in_var.0).clone())),
+                }
+            };
+
+            match where_fn.binding {
+                Binding::BindScalar(_) => constant.into_unit_matrix(&lookup)?,
```

Comment:
I expect that a single value will be the most common use of `ground`. E.g.,

https://gist.github.com/vvvvalvalval/5547d8b46414b955c88f#file-datomic_datalog_count-related-entities-with-zero-sums-clj-L27

It seems that you've effectively translated the tuple form

`(ground [1 2] [?x ?y])`

to be equivalent to

`(ground 1 ?x) (ground 2 ?y)`

which is correct.